[ARM] Lower unaligned loads/stores to aeabi functions.

simpal01 opened the pull request:
When targeting architectures that do not support unaligned memory accesses or when explictly pass -mno-unaligned-access, it requires the compiler to expand each unaligned load/store into an inline sequences. For 32-bit operations this typically involves:

	1. 4× LDRB (or 2× LDRH),
	2. multiple shift/or instructions

These sequences are emitted at every unaligned access site, and therefore contribute significant code size in workloads that touch packed or misaligned structures.

When compiling with -Oz and in combination with -mno-unaligned-access, this patch lowers unaligned 32 bit and 64 bit loads and stores to below AEABI heper calls:
```
         __aeabi_uread4
	__aeabi_uread8
	__aeabi_uwrite4
	__aeabi_uwrite8
```

And it provide a way to perform unaligned memory accesses on targets that do not support them, such as ARMv6-M or when compiling with -mno-unaligned-access. Although each use introduces a function call making it less straightforward than using raw loads and stores the call itself is often much smaller than the compiler emitted sequence of multiple ldrb/strb operations. As a result, these helpers can greatly reduce code-size providing they are invoked more than once across a program.

1. Functions become smaller in AEABI mode once they contain more than a few unaligned accesses.
2. The total image .text size becomes smaller whenever multiple functions call the same helpers.

This PR is derived from https://reviews.llvm.org/D57595, with some minor changes.
 Co-authored-by: David Green

simpal01 committed:
[ARM] Lower unaligned loads/stores to aeabi functions.

When targeting architectures that do not support unaligned
memory accesses or when explictly pass -mno-unaligned-access,
it requires the compiler to expand each unaligned load/store
into an inline sequences. For 32-bit operations this
typically involves:

	1. 4× LDRB (or 2× LDRH),
	2. multiple shift/or instructions

These sequences are emitted at every unaligned access site, and
therefore contribute significant code size in workloads that touch
packed or misaligned structures.

When compiling with -Os or -Oz in combination with
-mno-unaligned-access, this patch lowers unaligned 32 bit and 64 bit
loads and stores to below AEABI heper calls:
        aeabi_uread4
	aeabi_uread8
	aeabi_uwrite4
	aeabi_uwrite8

And it provide a way to perform unaligned memory accesses on
targets that do not support them, such as ARMv6-M or when
compiling with -mno-unaligned-access. Although each use
introduces a function call making it less straightforward
than using raw loads and stores the call itself is often
much smaller than the compiler emitted sequence of multiple
ldrb/strb operations. As a result, these helpers can greatly
reduce code-size providing they are invoked more than
once across a program.

1. Functions become smaller in AEABI mode once they contain more
   than a few unaligned accesses.
2. The total image .text size becomes smaller whenever multiple
   functions call the same helpers.

This PR is derived from https://reviews.llvm.org/D57595, with additional changes.
Co-authored-by: David Green

simpal01 committed:
fixup! [ARM] Lower unaligned loads/stores to aeabi functions.

1. Updated implementation to route all libcall handling
through RuntimeLibcalls, as recommended.
2. Logic has been updated so the behaviour now applies
   only for -Oz, not for both -Os and -Oz.

simpal01 commented on llvm/lib/Target/ARM/ARMISelLowering.cpp:
Done. Updated implementation to route all libcall handling through RuntimeLibcalls, as recommended.

efriedma-quic commented on llvm/lib/Target/ARM/ARMISelLowering.cpp:
Extra parentheses

efriedma-quic commented on llvm/test/CodeGen/ARM/unaligned_load_store_aeabi.ll:
Please make sure we have test coverage for optsize without minsize.

simpal01 committed:
fixup! [ARM] Lower unaligned loads/stores to aeabi functions.

   1. Removed extra parentheses.
   2. Test unaligned_load_store_aeabi.ll is updated
      to have coverage for both optsize and minsize cases.
   3. Forwards the chanis as well in the makelibcall
      to ensure correct scheduling and ordering.

simpal01 committed:
fixup! [ARM] Lower unaligned loads/stores to aeabi functions.

Fix clang-format issues.

simpal01 committed:
fixup! [ARM] Lower unaligned loads/stores to aeabi functions.

Removing extra Parentheses.

efriedma-quic reviewed (approved):
LGTM
